nav2_assisted_teleop/src/nav2_assisted_teleop.cpp

```diff
@@ -0,0 +1,311 @@
+      true);
+
+    vel_pub_ = create_publisher<geometry_msgs::msg::Twist>(output_vel_topic_, 1);
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  AssistedTeleopServer::on_activate(const rclcpp_lifecycle::State &)
+  {
+    RCLCPP_INFO(logger_, "Activating assisted_teleop_server");
+    vel_pub_->on_activate();
+    action_server_->activate();
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  AssistedTeleopServer::on_deactivate(const rclcpp_lifecycle::State &)
+  {
+    RCLCPP_INFO(logger_, "Deactivating assisted_teleop_server");
+
+    vel_pub_->on_deactivate();
+    action_server_->deactivate();
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  AssistedTeleopServer::on_cleanup(const rclcpp_lifecycle::State &)
+  {
+    RCLCPP_INFO(logger_, "Cleaning up assisted_teleop_server");
+
+    vel_pub_->on_activate();
+    action_server_->activate();
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  void AssistedTeleopServer::adjustVelocityIfInvalid()
+  {
+    vel_sub_ = create_subscription<geometry_msgs::msg::Twist>(
+    input_vel_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleopServer::inputVelocityCallback,
+      this, std::placeholders::_1));
+
+    joy_sub_ = create_subscription<sensor_msgs::msg::Joy>(
+    joystick_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleopServer::joyCallback,
+      this, std::placeholders::_1));
+
+    auto start_time = steady_clock_.now();
+
+    auto result = std::make_shared<Action::Result>();
+    rclcpp::WallRate loop_rate(10);
+    while (rclcpp::ok())
+    {
+      RCLCPP_INFO(logger_, "Running");
+      // user states that teleop was successful
+      if( joy_.buttons.size() > 0 && joy_.buttons[0] > 0)
+      {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
+        vel_sub_.reset();
+        joy_sub_.reset();
+        result->total_elapsed_time = steady_clock_.now() - start_time;
+        action_server_->succeeded_current(result);
+        return;
+      }
+
+      // user states that teleop failed
+      if( joy_.buttons.size() > 1 && joy_.buttons[1] > 0)
+      {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
+        vel_sub_.reset();
+        joy_sub_.reset();
+        result->total_elapsed_time = steady_clock_.now() - start_time;
+        action_server_->terminate_all(result);
+        return;
+      }
+
+      // cancel request
+      if(action_server_->is_cancel_requested())
+      {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
+        vel_sub_.reset();
+        joy_sub_.reset();
+        result->total_elapsed_time = steady_clock_.now() -start_time;
+        action_server_->terminate_all(result);
+        return;
+      }
+
+      // preempt request
+      if (action_server_->is_preempt_requested()) {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
```

Comment:
Shouldn't stop if preempted

Reply:
on stop if canceled, preempt implies that it has a new command so it can be fluidly transitioned to the next goal